### DISCUSSION TITLE
WebSockets

### DISCUSSION THREAD

**Discussion by jvo203** [zigzap/zap#6] [Ideas]
Hi, I know you're busy with other things but, time permitting, it would be nice to have the _server-side_ **WebSockets** support with a corresponding example program added to zap. Originally ```facil.io``` contains the WebSockets functionality too.

**Comment by renerocksai** [comment]
Hi, great idea! Time really is the limiting factor here. My first big use-case doesn't require websockets in any shape or form, just a good old REST API; however, part of the reason for zap is that I want to re-introduce multi-player mode to my latest iteration of a slideshow program. So I plan to use websockets there. ETA is ... summer.

If you or anyone you might know wants to explore adding / exporting websockets functionality to zap so we might get there earlier, I am open to PRs, though.

**Reply by jvo203** [reply]
Thanks, it's good to hear that. A bit of a background: I'm new to facil.io C API, right now am using the C libmicrohttpd for the HTTP part in combination with the C mongoose networking library for the HTTP/WebSockets part in the existing C application. At some point I would like to try to move the entire C application to ZIG just to see if there are any advantages (kick the tyres, see if ZIG fulfills all the promises etc.).

In the meantime, since ZIG is not yet production-ready, in the existing C application I am thinking of replacing the C libmicrohttpd / mongoose combo by a single C framework: facil.io. There will be a bit of a learning curve to get hang of facil.io. Should the transition to facil.io go well in the existing C code, the time might come to make a transition to ZIG and ZAP (with WebSockets).

So I am definitely keeping a keen eye on ZAP, hoping the WebSockets bindings might come at some point. Let's see how things go, depending on the timing you will probably beat me to it. I might become an early tester of ZAP with WebSockets. Or perhaps even contribute a Pull Request with WebSockets bindings. But first thing first, it is time to get a hang of the facil.io framework in C before learning ZIG and then ZAP.

**Comment by basilrwx** [comment]
I don't have the expertise to add/port ws from facil.io to zap. But I am really interested in using it through zap. Just here to express interest.

**Comment by renerocksai** [comment]
Just an update. I looked at it yesterday "by accident" while browsing the facil.io docs. It seems trivial (facil.io does all the work). So I'll add it soon. At the moment, my sycl talk and my work project are blocking me.

**Comment by renerocksai** [comment]
Websockets + example is in 00291117d85144fe62c394afa1ebc634298ec7f4 :smile:

**Comment by jvo203** [comment]
That's great!
